samples/server.c

```diff
@@ -224,6 +224,7 @@ static void do_dma_io(vfu_ctx_t *vfu_ctx, struct server_data *server_data)
         if (ret < 0) {
             err(EXIT_FAILURE, "vfu_sgl_write failed");
         }
+        vfu_sgl_mark_dirty(vfu_ctx, sg, 1);
```

Comment:
I'm not convinced this should be the server's responsibility; shouldn't we be marking dirty in vfu_sgl_write() ?

Reply:
I think we had to make marking dirty explicit because of implications related to quiesce/migration (don't remember the details), PRs #672 and #675 are related.

Reply:
I think usually if we use `vfu_sgl_get/put` it's done automatically but because we're using `vfu_sgl_write` we have to do it manually?

Reply:
@tmakatos I don't see how vfu_sgl_write() is any different from vfu_sgl_put() in this situation

Reply:
> I'm not convinced this should be the server's responsibility; shouldn't we be marking dirty in vfu_sgl_write() ?

`vfu_sgl_write()` doesn't mark the page dirty. This might make sense (I'm not sure it was intentional to begin with), as the vfio-user client knows this, it handles the DMA write itself as it's an explicit message. However, it's not consistent with the expectation that the server must explicitly mark the page dirty when it directly modifies the memory-mapped guest RAM.

I see pros and cons in both cases, @jlevon thoughts?

In any case, I think this part of the code is not fit for purpose as it tries to do two things:

1. demonstrate the use of `vfu_sgl_write`
2. mark guest RAM dirty during migration (via `vfu_sgl_get`/`vfu_sgl_mark_dirty`).

@w-henderson if you do the above then you should be getting dirty pages because of (2), whether or not you need to adapt the sample to consider pages dirtied by (1) will depend what we decide about `VFIO_USER_DMA_WRITE` (whether it should mark and reported pages dirtied by it).

Reply:
> vfu_sgl_write() doesn't mark the page dirty

I'm arguing it should be doing so. vfu_sgl_mark_dirty() exists for consumers that don't always want to vfu_sgl_put() for long-running mappings. That's not relevant for vfu_sgl_write().

Reply:
I think I probably agree with John that `vfu_sgl_write` should be marking the page dirty - I can't think of any situation in which you would want to be logging dirty pages for a certain region and then intentionally not log the dirtying of a page in that region...

Reply:
> > In any case, I think this part of the code is not fit for purpose as it tries to do two things:
>
> I'm arguing it should be doing so.

My comment is no longer valid therefore the only fix required is for `vfu_sgl_write` to call `vfu_sgl_mark_dirty`.

Reply:
My interpretation of [John's comments](https://github.com/nutanix/libvfio-user/pull/753#discussion_r1272506888) was the client does the dirty page tracking in case of `VFIO_USER_DMA_WRITE`, so the server shouldn't have to call `vfu_sgl_mark_dirty` at all. To demonstrate dirty page tracking, you'll have to do what in my previous comment (which I said that was no longer valid, but now it is).